modules/transfers/src/domain/transfers_agg.ts

```diff
@@ -11,31 +11,42 @@ import { TransfersFactory } from './transfers_factory'
 import { AckPayerFundsReservedCmd } from '../messages/ack_payer_funds_reserved_cmd'
 import { FulfilTransferCmd } from '../messages/fulfil_transfer_cmd'
 import { AckPayeeFundsCommittedCmd } from '../messages/ack_payee_funds_committed_cmd'
+import { IDupTransferRepo } from './transfers_duplicate_repo'
 
 export enum TransfersAggTopics {
   'Commands' = 'TransferCommands',
   'DomainEvents' = 'TransferDomainEvents',
 }
 
 export class TransfersAgg extends BaseAggregate<TransferEntity, TransferState> {
-  constructor (entityStateRepo: IEntityStateRepository<TransferState>, msgPublisher: IMessagePublisher, logger: ILogger) {
+  _duplicateRepo: IDupTransferRepo
+
+  constructor (entityStateRepo: IEntityStateRepository<TransferState>, duplicateRepo: IDupTransferRepo, msgPublisher: IMessagePublisher, logger: ILogger) {
     super(TransfersFactory.GetInstance(), entityStateRepo, msgPublisher, logger)
     this._registerCommandHandler('PrepareTransferCmd', this.processPrepareTransferCommand)
     this._registerCommandHandler('AckPayerFundsReservedCmd', this.processAckPayerFundsReservedCommand)
     this._registerCommandHandler('AckPayeeFundsCommittedCmd', this.processAckPayeeFundsReservedCommand)
     this._registerCommandHandler('FulfilTransferCmd', this.processFulfilTransferCommand)
+    this._duplicateRepo = duplicateRepo
   }
 
   async processPrepareTransferCommand (commandMsg: PrepareTransferCmd): Promise<boolean> {
     // try loading first to detect duplicates
 
-    await this.load(commandMsg.payload.transferId, false)
+    const isTransferAddedToDuplicateRepo = await this._duplicateRepo.add(commandMsg.payload.transferId)
```

Comment:
In here we should only check if exists,, the add should be at then end. Otherwise if the command processing fails after this line we end up with a fake duplicate.

Reply:
We are currently doing a similar duplicate check in the existing code-base due to the safety concerns. If we do not wish this to happen, we should have compensating actions to remove it from the duplicate set if the message fails to process, and it should only be removed from the duplicate store in specific use-cases.